---	
title: "Analyse mutation calls for DLBCL exomes"	
author: "Stacy Hung"	
date: "September 12, 2019"	
output: html_document	
---	
	
This script performs the following tasks:	
- reads in master file snv/indel results (summarized by canonical transcript and including batch and tumor content)
- plots distribution of tumor allele frequency per tumor sample, with information on tumor content
- plots #s of mutations per sample, stacked by variant class
- plots mutation frequency by panel gene (across total cohort) incl. non-mutated genes
 	
## Load libraries
 	
```{r}
library(dplyr)    # filter and rename
library(tidyr)    # separate	
```

## Plot total numbers of mutations stacked by variant class

```{r}
library(ggplot2)
library(plyr) # revalue function

# make sure mutations are unique (based on a single transcript)
calls.all <- read.table("/Volumes/shung/projects/AW_DLBCL_exomes_2019/calls.integration/snvs.all.intersect.with_germline_filters.txt", sep = "\t", header = TRUE, fill = TRUE)

calls.all$class <- revalue(calls.all$effect, c("NON_SYNONYMOUS_CODING"="Missense", 
                           "NON_SYNONYMOUS_START"="Missense",
                           "START_GAINED"="Missense",
                           "START_LOST"="Start Lost",
                           "STOP_LOST"="Missense",
                           "STOP_GAINED"="Nonsense",
                           "STOP_GAINED+CODON_CHANGE_PLUS_CODON_DELETION"="Nonsense",
                           "CODON_CHANGE_PLUS_CODON_DELETION"="In-frame Indel",
                           "CODON_CHANGE_PLUS_CODON_INSERTION"="In-frame Indel",
                           "CODON_DELETION"="In-frame Indel",
                           "CODON_INSERTION"="In-frame Indel",
                           "FRAME_SHIFT"="Frameshift Indel",
                           "FRAME_SHIFT+START_LOST"="Frameshift Indel",
                           "SPLICE_SITE_ACCEPTOR"="Splice site",
                           "SPLICE_SITE_DONOR"="Splice site",
                           "UTR_3_PRIME"="3' UTR",
                           "UTR_5_PRIME"="5' UTR"
                           ))


calls.all$tumor_id <- factor(calls.all$tumor_id)
data.df <- as.data.frame(table(calls.all$tumor_id, calls.all$class))
colnames(data.df) <- c("case", "variant_class", "count")

# sort by total number of mutations
p <- ggplot(data = data.df, aes(x = reorder(case, -count), y = count, fill = variant_class)) + 
  geom_bar(stat = "identity", width = 0.6) + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  scale_fill_discrete(name = "Variant Class") +
  xlab("") + ylab("Number of mutations") 
```

