---	
title: "Extract mutation calls for DLBCL genomes"	
author: "Stacy Hung"	
date: "November 12, 2019"	
output: html_document	
---	
	
This script performs the following tasks:	
- reads variants as published in Arthur et al. Nat. Comm 2018 - file has been filtered to exclude intronic, intergenic and 3'/5' flanking variants
- extracts calls in 18 genomes of interest

## Load libraries
 	
```{r}
library(dplyr)    # filter and rename
library(tidyr)    # separate	
library(data.table) # for renaming columns
```

## Read in datasets of interest

```{r}
genome_cases <- read.table("/Volumes/shung/projects/AW_DLBCL_exomes_and_genomes_2019/genomes/data/genomes-ids.txt", sep = "\t", header = TRUE, fill = TRUE)

# original set of calls
genome_calls.p1 <- read.table("/Volumes/shung/projects/AW_DLBCL_exomes_and_genomes_2019/genomes/Arthur_et_al-Nat_Comm-2018/part1/AW_gph_genomes.filtered.minus_missing_anno_cases.min_columns.txt", sep = "\t", header = TRUE, fill = TRUE)

# calls for silent mutations in original set of calls
genome_calls.silent.p1 <- read.table("/Volumes/shung/projects/AW_DLBCL_exomes_and_genomes_2019/genomes/Arthur_et_al-Nat_Comm-2018/part1/AW_gph_genomes.silent.minus_missing_anno_cases.min_col.txt", sep = "\t", header = TRUE, fill = TRUE)

# calls for cases that were missing tumor read and allele frequency information
# specifically, these were the cases:
#     14-35026
#     15-24058
#     15-24306
genome_calls.missing_cases <- read.table("/Volumes/shung/projects/AW_DLBCL_exomes_and_genomes_2019/genomes/Arthur_et_al-Nat_Comm-2018/missing_annotation_from_part1/AW_gph_genomes.missing_annotations.filtered.min_col.txt", sep = "\t", header = TRUE, fill = TRUE)

# extract min columns and create a single genome calls dataset
keeps <- c("Hugo_Symbol", "NCBI_Build", "Chromosome", "Start_Position", "End_Position", "Variant_Classification",
           "Variant_Type", "Reference_Allele", "Tumor_Seq_Allele1", "Tumor_Seq_Allele2", "dbSNP_RS", "Tumor_Sample_Barcode",
           "Matched_Norm_Sample_Barcode", "Match_Norm_Seq_Allele1", "Match_Norm_Seq_Allele2", "HGVSc",
           "HGVSp_Short", "Transcript_ID", "t_depth", "t_ref_count", "t_alt_count", "n_depth", "n_ref_count",
           "n_alt_count", "Allele", "Gene", "Existing_variation", "BIOTYPE", "IMPACT")

genome_calls.p1 <- genome_calls.p1[, keeps]
genome_calls.silent.p1 <- genome_calls.silent.p1[, keeps]
genome_calls.missing_cases <- genome_calls.missing_cases[, keeps]
genome_calls <- rbind (genome_calls.p1, genome_calls.silent.p1, genome_calls.missing_cases)
rm(genome_calls.p1)
rm(genome_calls.silent.p1)
rm(genome_calls.missing_cases)

# remove "RNA" variant classification
genome_calls <- filter(genome_calls, genome_calls$Variant_Classification != "RNA")
genome_calls$Variant_Classification <- factor(genome_calls$Variant_Classification)

write.table(genome_calls, "/Volumes/shung/projects/AW_DLBCL_exomes_and_genomes_2019/genomes/gph_genomes.mutations.filtered.combined_missing_annotation.txt", sep = "\t", quote = FALSE, row.names = FALSE)

exome_calls <- read.table("/Volumes/shung/projects/AW_DLBCL_exomes_and_genomes_2019/exomes/maf/all.AW_DLBCL_exomes_2019.mutations.maf", sep = "\t", header = TRUE, fill = TRUE)
```

## Filter genome calls

```{r}
# rename columns
setnames(genome_calls, "Existing_variation", "External_ID")
setnames(genome_calls, "HGVSp_Short", "Protein_Change")
setnames(genome_calls, "HGVSc", "CDS_Change")
setnames(genome_calls, "t_alt_count", "Tumor_Var_Reads")

# calculate VAF column
#genome_calls <- genome_calls %>%
#  mutate(Tumor_Allele_Freq = ifelse(as.numeric(genome_calls$t_depth) > 0, 
#                      (as.numeric(genome_calls$t_alt_count) / as.numeric(genome_calls$t_depth))*100, 
#                      "NA"))
genome_calls$Tumor_Allele_Freq <- as.numeric(
  (as.numeric(genome_calls$Tumor_Var_Reads) / as.numeric(genome_calls$t_depth))*100
  )

# extract only columns of interest (and reorder)
keeps <- c("Tumor_Sample_Barcode", "Variant_Type", "Hugo_Symbol", 
           "Chromosome", "Start_Position", "End_Position", 
           "Reference_Allele", "Tumor_Seq_Allele1", "Tumor_Seq_Allele2", 
           "Tumor_Var_Reads", "Tumor_Allele_Freq",
           "Variant_Classification", "Protein_Change", "CDS_Change", "External_ID")
genome_calls <- genome_calls[keeps]

# output bed column to be used for hgLiftover from hg38 to hg19 (since exome calls are based on hg19)
genome_calls$bed <- as.character(
  paste(paste(genome_calls$Chromosome,genome_calls$Start_Position,sep = ":"), genome_calls$End_Position, sep = "-")
  )
write.table(genome_calls, "/Volumes/shung/projects/AW_DLBCL_exomes_and_genomes_2019/genomes/gph_genomes.mutations.filtered.with_bed.txt", sep = "\t", row.names = FALSE, quote = FALSE)

#write.table(genome_calls, "/Volumes/shung/projects/AW_DLBCL_exomes_and_genomes_2019/genomes/gph_genomes.mutations.filtered.maf", sep = "\t", row.names = FALSE, quote = FALSE)
```

## Create hg19 version of genome mutation calls

```{r}
# read in list of calls with hg38 coordinates
genome_calls <- read.table("/Volumes/shung/projects/AW_DLBCL_exomes_and_genomes_2019/genomes/gph_genomes.mutations.filtered.maf", sep = "\t", header = TRUE, fill = TRUE)

# read in liftover coordinates (hg19)
genome_calls.with_hg19_bed <- read.table("/Volumes/shung/projects/AW_DLBCL_exomes_and_genomes_2019/genomes/hglft_genome_32433_b50060.bed")
colnames(genome_calls.with_hg19_bed) <- c("hg19_liftover")

# combine
genome_calls <- cbind(genome_calls, genome_calls.with_hg19_bed)

# replace bed column with hg19 coords
temp <- data.frame(do.call('rbind', strsplit(as.character(genome_calls$hg19_liftover),':',fixed=TRUE)))
colnames(temp) <- c("chr.hg19", "start-end")
temp2 <- data.frame(do.call('rbind', strsplit(as.character(temp$`start-end`),'-',fixed=TRUE)))
colnames(temp2) <- c("Start.hg19", "End.hg19")
genome_calls <- cbind(genome_calls, temp$chr.hg19, temp2$Start.hg19, temp2$End.hg19)
genome_calls$Chromosome <- NULL
genome_calls$Start_Position <- NULL
genome_calls$End_Position <- NULL
genome_calls$hg19_liftover <- NULL

colnames(genome_calls) <- c("Tumor_Sample_Barcode", "Variant_Type", "Hugo_Symbol", "Reference_Allele", "Tumor_Seq_Allele1", "Tumor_Seq_Allele2", "Tumor_Var_Reads", "Tumor_Allele_Freq", "Variant_Classification", "Protein_Change", "CDS_Change", "External_ID", "Chromosome", "Start_Position", "End_Position")

# place columns in same order as exome columns
genome_calls <- genome_calls[, colnames(exome_calls)]

# write to file
write.table(genome_calls, "/Volumes/shung/projects/AW_DLBCL_exomes_and_genomes_2019/genomes/gph_genomes.mutations.hg19.maf", sep = "\t", row.names = FALSE, quote = FALSE)

# combine with exome calls
exome_calls <- read.table("/Volumes/shung/projects/AW_DLBCL_exomes_and_genomes_2019/exomes/maf/all.AW_DLBCL_exomes_2019.mutations.maf", sep = "\t", header = TRUE, fill = TRUE)
calls.all <- rbind(genome_calls, exome_calls)

# write to file
write.table(calls.all, "/Volumes/shung/projects/AW_DLBCL_exomes_and_genomes_2019/AW_DLBLC_2019-snvs_and_indels.exomes_and_genomes.hg19.maf", sep = "\t", quote = FALSE, row.names = FALSE)

# create version specific for oncodriveclustl (newest version, via command line)
maf.cols <- colnames(calls.all)
oncodriveclust.cols <- c("SAMPLE", "VARIANT_TYPE", "GENE", "CHROMOSOME", "POSITION", "END_POSITION", "REF", 
                         "ALT_1", "ALT", "TUMOR.VAR_READS", "TUMOR.VAF", "VARIANT_CLASS", "HGVS_PROTEIN",
                         "HGVS_CDNA", "ID")
colnames(calls.all) <- oncodriveclust.cols
write.table(calls.all, "/Volumes/shung/projects/AW_DLBCL_exomes_and_genomes_2019/oncodriveclustl/input/AW_DLBLC_2019-snvs_and_indels.exomes_and_genomes.hg19.tsv", sep = "\t", quote = FALSE, row.names = FALSE)
```

## try oncoclust

```{r}
library(maftools)

calls.maf <- read.maf("/Volumes/shung/projects/AW_DLBCL_exomes_and_genomes_2019/AW_DLBLC_2019-snvs_and_indels.exomes_and_genomes.hg19.maf")

calls.sig <- oncodrive(maf = calls.maf, AACol = 'Protein_Change', minMut = 3, pvalMethod = 'zscore')
#calls.sig <- oncodrive(maf = calls.maf, AACol = 'Protein_Change', minMut = 5, pvalMethod = 'zscore', bgEstimate = FALSE)
head(calls.sig)

plotOncodrive(res = calls.sig, fdrCutOff = 0.1, useFraction = TRUE)


```

## combine exome and genome datasets

```{r}
# version of calls that should be used as input into MutSigCv - currently missing silent mutations from exome calls
calls.all <- rbind(exome_calls, genome_calls)

# version of calls to look for somatic mutations
calls.somatic <- filter(calls.all, calls.all$Variant_Classification != "Silent")

write.table(calls.all, "/Volumes/shung/projects/AW_DLBCL_exomes_and_genomes_2019/AW_DLBLC_2019-snvs_and_indels.exomes_and_genomes.txt", sep = "\t", row.names = FALSE, quote = FALSE)

```


## Plot total numbers of mutations stacked by variant class

```{r}
library(ggplot2)
library(plyr) # revalue function

# make sure mutations are unique (based on a single transcript)
calls.all <- read.table("/Volumes/shung/projects/AW_DLBCL_exomes_2019/calls.integration/snvs.all.intersect.with_germline_filters.txt", sep = "\t", header = TRUE, fill = TRUE)

calls.all$class <- revalue(calls.all$effect, c("NON_SYNONYMOUS_CODING"="Missense", 
                           "NON_SYNONYMOUS_START"="Missense",
                           "START_GAINED"="Missense",
                           "START_LOST"="Start Lost",
                           "STOP_LOST"="Missense",
                           "STOP_GAINED"="Nonsense",
                           "STOP_GAINED+CODON_CHANGE_PLUS_CODON_DELETION"="Nonsense",
                           "CODON_CHANGE_PLUS_CODON_DELETION"="In-frame Indel",
                           "CODON_CHANGE_PLUS_CODON_INSERTION"="In-frame Indel",
                           "CODON_DELETION"="In-frame Indel",
                           "CODON_INSERTION"="In-frame Indel",
                           "FRAME_SHIFT"="Frameshift Indel",
                           "FRAME_SHIFT+START_LOST"="Frameshift Indel",
                           "SPLICE_SITE_ACCEPTOR"="Splice site",
                           "SPLICE_SITE_DONOR"="Splice site",
                           "UTR_3_PRIME"="3' UTR",
                           "UTR_5_PRIME"="5' UTR"
                           ))


calls.all$tumor_id <- factor(calls.all$tumor_id)
data.df <- as.data.frame(table(calls.all$tumor_id, calls.all$class))
colnames(data.df) <- c("case", "variant_class", "count")

# sort by total number of mutations
p <- ggplot(data = data.df, aes(x = reorder(case, -count), y = count, fill = variant_class)) + 
  geom_bar(stat = "identity", width = 0.6) + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  scale_fill_discrete(name = "Variant Class") +
  xlab("") + ylab("Number of mutations") 
```

